diff --git a/src/p7_tophits.c b/src/p7_tophits.c
index cd2bf208..365ccbf0 100644
--- a/src/p7_tophits.c
+++ b/src/p7_tophits.c
@@ -2040,7 +2040,11 @@ p7_tophits_TabularTail(FILE *ofp, const char *progname, enum p7_pipemodes_e pipe
 
   if ((status = esl_opt_SpoofCmdline(go, &spoof_cmd)) != eslOK) goto ERROR;
   if (date == -1)                                               ESL_XEXCEPTION(eslESYS, "time() failed");
+#if defined(_WIN32) || defined(_WIN64) || defined(__MINGW32__) || defined(__CYGWIN__)
+  if (ctime_s(timestamp, sizeof(timestamp), &date) != 0)        ESL_XEXCEPTION(eslESYS, "ctime_s() failed");
+#else
   if ((ctime_r(&date, timestamp)) == NULL)                      ESL_XEXCEPTION(eslESYS, "ctime_r() failed");
+#endif
   switch (pipemode) {
     case p7_SEARCH_SEQS: strcpy(modestamp, "SEARCH"); break;
     case p7_SCAN_MODELS: strcpy(modestamp, "SCAN");   break;
